path_login.go

```diff
@@ -280,10 +280,57 @@ func (b *kubeAuthBackend) parseAndValidateJWT(jwtStr string, role *roleStorageEn
 // serviceAccount holds the metadata from the JWT token and is used to lookup
 // the JWT in the kubernetes API and compare the results.
 type serviceAccount struct {
-	Name       string `mapstructure:"kubernetes.io/serviceaccount/service-account.name"`
-	UID        string `mapstructure:"kubernetes.io/serviceaccount/service-account.uid"`
-	SecretName string `mapstructure:"kubernetes.io/serviceaccount/secret.name"`
-	Namespace  string `mapstructure:"kubernetes.io/serviceaccount/namespace"`
+	Name       string   `mapstructure:"kubernetes.io/serviceaccount/service-account.name"`
+	UID        string   `mapstructure:"kubernetes.io/serviceaccount/service-account.uid"`
+	SecretName string   `mapstructure:"kubernetes.io/serviceaccount/secret.name"`
+	Namespace  string   `mapstructure:"kubernetes.io/serviceaccount/namespace"`
+	Aud        []string `mapstructure:"aud"`
+
+	// the JSON returned from reviewing a Projected Service account has a
+	// different structure, where the information is in a sub-structure instead of
+	// at the top level
+	Kubernetes *projectedServiceToken `mapstructure:"kubernetes.io"`
+	Expiration int64                  `mapstructure:"exp"`
+	IssuedAt   int64                  `mapstructure:"iat"`
+}
+
+// uid returns the UID for the service account, preferring the projected service
+// account value if found
+func (s *serviceAccount) uid() string {
+	if s.Kubernetes != nil && s.Kubernetes.ServiceAccount != nil {
+		return s.Kubernetes.ServiceAccount.UID
+	}
+	return s.UID
+}
+
+// name returns the name for the service account, preferring the projected
+// service account value if found. This is "default" for projected service
+// accounts
+func (s *serviceAccount) name() string {
+	if s.Kubernetes != nil && s.Kubernetes.ServiceAccount != nil {
```

Comment:
Should this one have an additional check as well for whether the string itself is populated?

Reply:
Same reply as above